akka-persistence-testkit/src/main/scala/akka/persistence/testkit/state/scaladsl/PersistenceTestKitDurableStateStore.scala

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.testkit.state.scaladsl
+
+import java.util.concurrent.atomic.AtomicLong
+
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.Future
+
+import akka.Done
+import akka.actor.ExtendedActorSystem
+import akka.persistence.query.scaladsl.DurableStateStoreQuery
+import akka.persistence.query.DurableStateChange
+import akka.persistence.query.Offset
+import akka.persistence.query.NoOffset
+import akka.persistence.query.Sequence
+import akka.persistence.state.scaladsl.{ DurableStateUpdateStore, GetObjectResult }
+import akka.stream.scaladsl.BroadcastHub
+import akka.stream.scaladsl.Keep
+import akka.stream.scaladsl.Source
+import akka.stream.typed.scaladsl.ActorSource
+import akka.stream.OverflowStrategy
+
+object PersistenceTestKitDurableStateStore {
+  val Identifier = "akka.persistence.testkit.state"
+}
+
+class PersistenceTestKitDurableStateStore[A](val system: ExtendedActorSystem)
+    extends DurableStateUpdateStore[A]
+    with DurableStateStoreQuery[A] {
+
+  private implicit val sys = system
+  private val store = TrieMap.empty[String, Record[A]]
+
+  private val (publisher, changesSource) =
+    ActorSource
+      .actorRef[Record[A]](PartialFunction.empty, PartialFunction.empty, 256, OverflowStrategy.dropHead)
+      .toMat(BroadcastHub.sink)(Keep.both)
+      .run()
+
+  private val EarliestOffset = 0L
+  private val lastGlobalOffset = new AtomicLong(EarliestOffset)
+
+  def getObject(persistenceId: String): Future[GetObjectResult[A]] =
+    Future.successful(GetObjectResult(store.get(persistenceId).map(_.value), 0))
+
+  def upsertObject(persistenceId: String, revision: Long, value: A, tag: String): Future[Done] =
+    Future.successful {
+      val globalOffset = lastGlobalOffset.incrementAndGet()
+      val record = Record(globalOffset, persistenceId, revision, value, tag)
+      store.put(persistenceId, record)
+      publisher ! record
```

Comment:
This is not concurrency safe, might publish records in a different order than the global offset. We don't have to complicate this, since it's only intended for tests. Make all methods `synchronized`. Then the `TrieMap` can also changed to an ordinary immutable or mutable `Map`.

Reply:
yes I thought about that, I'll take the simple route, thanks.